ml/src/main/scala/org/apache/spark/ml/linalg/CUBLAS.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.linalg
+
+import com.nvidia.spark.ml.linalg.JniCUBLAS
+
+/**
+ * BLAS routines for MLlib's vectors and matrices.
+ */
+private[spark] object CUBLAS extends Serializable {
+
+  @transient private var _jniCUBLAS: JniCUBLAS = _
+
+  private[spark] def jniCUBLAS: JniCUBLAS = {
+    if (_jniCUBLAS == null) {
+      _jniCUBLAS = JniCUBLAS.getInstance
+    }
+    _jniCUBLAS
+  }
+
+  /**
+   * Adds alpha * v * v.t to a matrix in-place. This is the same as BLAS's ?SPR.
+   *
+   * @param U the upper triangular part of the matrix packed in an array (column major)
+   */
+  def spr(v: DenseVector, U: Array[Double]): Unit = {
+    jniCUBLAS.dspr(v.size, v.values, U)
+  }
+
+  /**
+   * C := B.transpose * B
+   *
+   * @param B the matrix B that will be left multiplied by its transpose. Size of m x n.
+   * @param C the resulting matrix C. Size of n x n.
+   */
+  def gemm(B: DenseMatrix, C: DenseMatrix, deviceID: Int): Unit = {
```

Comment:
We'd better add more comments to explain the parameters for all these functions.